cajita/src/Cajita_SparseDimPartitioner.hpp

```diff
@@ -0,0 +1,687 @@
+                        Kokkos::RangePolicy<execution_space>(
+                            0, _ranks_per_dim[dj] * _ranks_per_dim[dk] ),
+                        KOKKOS_LAMBDA( const int idx, int& update ) {
+                            update += current_workload( idx );
+                        },
+                        diff );
+                    // record the new optimal position
+                    if ( diff <= last_diff )
+                    {
+                        // check if point_i reach the total_tile_num
+                        if ( point_i == rec_mirror( rank, di ) )
+                        {
+                            rec_mirror( current_rank, di ) = point_i;
+                            break;
+                        }
+                        last_diff = diff;
+                        point_i++;
+                    }
+                    else
+                    {
+                        // final optimal position
+                        if ( rec_mirror( current_rank, di ) != point_i - 1 )
+                        {
+                            rec_mirror( current_rank, di ) = point_i - 1;
+                            is_changed = true;
+                        }
+                        last_point = point_i - 1;
+                        break;
+                    }
+                } // end while (optimization for the current rank)
+            }     // end for (all partition/rank in the optimized dimension)
+            Kokkos::deep_copy( _rectangle_partition_dev, rec_mirror );
+        } // end for (3 dimensions)
+    }
+
+    /*!
+      \brief compute the imbalance factor for the current partition
+      \param cart_comm MPI cartesian communicator
+      \return the imbalance factor = workload on current rank / ave_workload
+    */
+    float computeImbalanceFactor( MPI_Comm cart_comm )
+    {
+        auto rec_partition = _rectangle_partition_dev;
+        auto prefix_sum = _workload_prefix_sum;
+        // workload_record[0] = ave_workload
+        // workload_record[1] = current_workload
+        Kokkos::View<float[2], memory_space> workload_record(
+            "workload_record" );
+
+        // Get the Cartesian topology index of this rank.
+        Kokkos::Array<int, 3> cart_rank;
+        int linear_rank;
+        MPI_Comm_rank( cart_comm, &linear_rank );
+        MPI_Cart_coords( cart_comm, linear_rank, 3, cart_rank.data() );
+
+        SubWorkloadFunctor compute_sub_workload = { _rectangle_partition_dev,
+                                                    _workload_prefix_sum };
+
+        int size_0 = prefix_sum.extent( 0 );
+        int size_1 = prefix_sum.extent( 1 );
+        int size_2 = prefix_sum.extent( 2 );
+
+        int rank_0 = _ranks_per_dim[0];
+        int rank_1 = _ranks_per_dim[1];
+        int rank_2 = _ranks_per_dim[2];
+
+        Kokkos::parallel_for(
+            "compute_workload_record",
+            Kokkos::RangePolicy<execution_space>( 0, 2 ),
```

Comment:
Probably not worth a parallel kernel here. Could copy what you need to the host instead